### DISCUSSION TITLE
Remapping/altering column data on migrate?

### DISCUSSION THREAD

**Discussion by AdamBridges** [simolus3/drift#2490] [Q&A] [answered]
Hello,

So I want to migrate a column of data from `List<String>` to `List<MyCustomClass>` like so: 

```
class Entry extends Table {

  IntColumn get id => integer().autoIncrement().nullable()();
  IntColumn get dateMilli => integer()();
  // TextColumn get data => text().map(
  //     StringToListConverter()).nullable()();
  TextColumn get data => text().map(
      StringToDataConverter()).nullable()();

}
```

However, I am uncertain how to do this exactly. I have the new `TypeConverter` properly setup for `List<MyCustomClass>` and I think I use the `alterTable()` function with `TableMigration()` when upgrading the schema, but I'm uncertain how to convert the old values with the `columnTransformer`. 

I've reviewed the documentation but due to the complexity of the data, I need to be able to remap values, so unless I've misunderstood something, I don't think a simple cast is going to do it as suggested:
```
columnTransformer: {
  todos.category: todos.category.cast<Type>(),
}
```

The only alternative approaches I've considered so far is using `customSelect()` from the table's DAO to get the old values to update with; or to add an entirely new column to migrate with. However, if I can avoid these alternatives with the use of something already built-in, then I'd rather learn how to make it happen. 

Any help will be appreciated.

Thanks!

**Comment by simolus3** [comment, marked as answer]
Just to be sure, the format of data in the database has changed as well, right? In other words, the type converter is unable to read existing before the migration? If it were able to do that, you wouldn't need a migration for this at all.

The `columnTransformer` API can only be used for conversions that can be expressed in SQL. The easiest solution that comes it mind is to use a custom function:

```dart
final db = Database(
  NativeDatabase.createInBackground(File('/tmp/foo'), setup: (rawDatabase) {
    rawDatabase.createFunction(
      functionName: 'migrate_entry_data',
      function: (args) {
        final oldData = args[0] as String;
        // This function needs to convert the entry.data column of an existing row into the new format
        return 'transformed row';
      },
      argumentCount: const AllowedArgumentCount(1),
    );
  }),
);
```

Then you can use something like

```dart
columnTransformer: {
  entry.data: FunctionCallExpression('migrate_entry_data', [entry.data])
}
```

**Reply by AdamBridges** [reply]
> Just to be sure, the format of data in the database has changed as well, right? In other words, the type converter is unable to read existing before the migration?

Yes, that is correct.

Thanks for the timely response. I really appreciate it. I'll try this out and see how it goes.

**Reply by AdamBridges** [reply]
This worked well. Thanks again! 

One point of confusion I initially had was where to create the function so for others who may have this issue, simply just create it within the call to the database that contains the table you're altering.

**Comment by simolus3** [comment]
Actually, it can be done a bit simpler. Since the actual schema of the table doesn't change (the data column stays a string as far as the database is concerned), you can write the schema migration with a select + update.

You'd still want to bump your schema version getter, but the migration could either be a simple update if using the custom function:

```dart
await entry.update().write(EntryCompanion.custom(
  id: FunctionCallExpression('migrate_entry_data', [entry.data])
));
```

If you don't want to register a custom function, you could load the old data (but you'd have to use a `customSelect` since a regular `select` would apply the new type converter), do the update in Dart, and then write the new rows back to the database.
